lib/pengine/rules.c

```diff
@@ -569,39 +569,33 @@ accept_attr_expr(const char *l_val, const char *r_val, enum pcmk__type type,
  * \internal
  * \brief Get correct value according to \c PCMK_XA_VALUE_SOURCE
  *
- * \param[in] expr_id       Rule expression ID (for logging only)
- * \param[in] value         value given in rule expression
- * \param[in] value_source  \c PCMK_XA_VALUE_SOURCE given in rule expressions
- * \param[in] match_data    If not NULL, resource back-references and params
+ * \param[in] value       value given in rule expression
+ * \param[in] source      Reference value source
+ * \param[in] match_data  If not NULL, resource back-references and params
  */
 static const char *
-expand_value_source(const char *expr_id, const char *value,
-                    const char *value_source, const pe_match_data_t *match_data)
+expand_value_source(const char *value, enum pcmk__reference_source source,
+                    const pe_match_data_t *match_data)
 {
     GHashTable *table = NULL;
 
-    if (pcmk__str_empty(value)) {
-        return NULL; // value_source is irrelevant
+    switch (source) {
+        case pcmk__source_literal:
+            return value;
 
-    } else if (pcmk__str_eq(value_source, PCMK_VALUE_PARAM, pcmk__str_casei)) {
-        table = match_data->params;
-
-    } else if (pcmk__str_eq(value_source, PCMK_VALUE_META, pcmk__str_casei)) {
-        table = match_data->meta;
+        case pcmk__source_instance_attrs:
+            table = match_data->params;
+            break;
 
-    } else { // literal
-        if (!pcmk__str_eq(value_source, PCMK_VALUE_LITERAL,
-                          pcmk__str_null_matches|pcmk__str_casei)) {
+        case pcmk__source_meta_attrs:
+            table = match_data->meta;
+            break;
 
-            pcmk__config_warn("Expression %s has invalid " PCMK_XA_VALUE_SOURCE
-                              " value '%s', using default "
-                              "('" PCMK_VALUE_LITERAL "')",
-                              pcmk__s(expr_id, "without ID"), value_source);
-        }
-        return value;
+        default:
+            return NULL; // Not possible
     }
 
-    if (table == NULL) {
+    if ((table == NULL) || pcmk__str_empty(value)) {
```

Comment:
This is a slight behavioral change. Previously, if `value` was an empty string and `source` was `PCMK_VALUE_LITERAL`, we'd return `NULL`. Now, we'll return the empty string in that case.

If we want to continue returning `NULL`, just move the `pcmk__str_empty()` check above the switch.

Reply:
Hmm, we should probably never map empty string to NULL for this purpose, but we need to preserve behavior for now